### PR TITLE
Add config.horizontal_scroll_list to enable horizontal scrolling colu…

### PR DESCRIPTION
**Pull request by sbull**
…mns and frozen row headers instead of paginated sets

Addresses Issue #2272.

### CODE REVIEW

File: app/views/rails_admin/main/index.html.haml [hunk 1]
```diff
@@ -39,11 +51,43 @@
...
+            if (idx === 0) {
+              firstPosition = $td.position().left;
+            }
+            td.style.left = ($td.position().left-firstPosition)+'px';
+          });
+        });
+      };
+      $(window).on('load', setFrozenColPositions); // Update after link icons load.
+      $(document).on('rails_admin.dom_ready', setFrozenColPositions);
+    }());
```
Comment (mshibuya) [inline app/views/rails_admin/main/index.html.haml]: Place your javascripts and stylesheets in appropriate location, not putting them into the view.

File: lib/rails_admin/config.rb [hunk 2]
```diff
@@ -59,6 +59,9 @@ class << self
       # Set the max width of columns in list view before a new set is created
       attr_accessor :total_columns_width
 
+      # Enable horizontal-scroll table in list view, ignore total_columns_width
+      attr_accessor :horizontal_scroll_list
```
Comment (mshibuya) [inline lib/rails_admin/config.rb]: Isn't it nice to being able to configure this per model basis?

File: app/views/rails_admin/main/index.html.haml [hunk 3]
```diff
@@ -12,10 +12,22 @@
...
+    end
+    horiz_scroll_frozen_cols ||= 3 # by default, freeze checkboxes, links & first property (usually id?)
+    horiz_scroll_frozen_cols -= 1 unless checkboxes
+  else
+    horiz_scroll_frozen_cols = 0
+    sets = get_column_sets(properties)
+    properties = sets[params[:set].to_i] || []
+    other_left = ((params[:set].to_i - 1) >= 0) && sets[params[:set].to_i - 1].present?
+    other_right = sets[params[:set].to_i + 1].present?
+  end
```
Comment (mshibuya) [inline app/views/rails_admin/main/index.html.haml]: Can these calculation logics be more simple?
If they needs to be that complex, consider moving it to outside of the view.

**Comment by sbull** [comment]
Thanks for your feedback!

1. I had wanted to make this as low-impact as possible, so that the JS & CSS wouldn't even be loaded in the default configuration. If these are in separate files, I don't think it's possible to conditionally load them (assuming the right thing would be to make them part of the asset precompile pipeline)? But I can submit a change to do it the "right" way.

2. I expect that the most-used config will be global options, but I will look into making it work per-model as well.

3. I'll look into moving some of the calculation logic into the config class or helpers.

I hope to put some time into this in a few days. Thanks again!

**Comment by sbull** [comment]
Hi @mshibuya, I've refactored this code to address your requests
1. Move JS & CSS into asset files.
2. Enable configuration at model level.
3. Move config logic out of view.

I've reworked the tests to account for these changes, and created new tests to check the model-level configuration. I've also rebased this PR. I can squash these commits if you'd like but I thought I'd leave them separate for now for you to review the changes from the original PR.

Let me know if other changes are needed. I don't really know how to fix the codeclimate "cognitive complexity" issue - perhaps split it up into multiple methods, but that seems like it could result in a fair amount of duplicate code.

Thanks!

**Comment by sbull** [comment]
Hi @mshibuya , I was wondering if you could take a look at this, as it's been open for a while. Thanks!

**Comment by mshibuya** [comment]
Thanks you for amazing work!
I'm planning to change the name of this feature to `sidescroll`.

**Comment by twikah** [comment]
Hi @mshibuya, thanks for getting this merged! Are you planning on making a release soon? It would be great!

**Pull request merged**